src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -119,21 +139,58 @@ async Task EnsureLong()
 
                 try
                 {
+                    var timeoutCt = CancellationToken.None;
+                    if (async)
+                    {
+                        if (Timeout > TimeSpan.Zero)
+                        {
+                            // We reuse the timeout's cancellation token source as long as it hasn't fired, but once it has
+                            // there's no way to reset it (see https://github.com/dotnet/runtime/issues/4694)
+                            _timeoutCts.CancelAfter(Timeout);
+                            if (_timeoutCts.IsCancellationRequested)
+                            {
+                                _timeoutCts.Dispose();
+                                _timeoutCts = new CancellationTokenSource(Timeout);
+                            }
+                            timeoutCt = _timeoutCts.Token;
+                        }
+                    }
+                    else
+                    {
+                        if (Timeout != _currentSocketTimeout && UnderlyingSocket != null)
```

Comment:
Sadly, we cannot move `null` check in the constructor, as there are cases, where having `null` for `UnderlyingSocket` makes sense (described above). But there are also some points to think about:

- Every time we do not set `UnderlyingSocket` (for example, using `MemoryStream` as `UnderlyingStream`), we also do not set timeout. This means that we can safely omit `null` check for `UnderlyingSocket`.
- But if this behavior ever changes, we will get `NullReferenceException`, which is definitely not user friendly.
- While `NpgsqlReadBuffer` is public (it seems to be so only for tests), it's not a part of public api.

Same goes for `NpgsqlWriteBuffer`.

Reply:
I agree with everything you wrote.

An Idea:
Since `Timeout`, `_currentSocketTimeout` and `UnderlyingSocket` are heavily interdependent, how about making `Timeout` a property backed by `_currentSocketTimeout` (or just call it `_timeout` then) and moving the process of setting `UnderlyingSocket.ReceiveTimeout` to the property setter.

This way we could even do a null check on `UnderlyingSocket` before setting it's `ReceiveTimeout` property and throw an `InvalidOperationException` if someone tries to set a timeout and (semi-)illegally passed `null!` for the socket to our constructor.

This way we would not only move the null check out of the hot path but everything that is associated with setting the socket timeout.

Reply:
After sleeping over it, I think we can even have two constructors.
One with socket and timeout and one without.
This way both can become private fields and we just don't support resetting the timeout after creating the buffer.
I don't see a reason why the option to reset it on an existing buffer would be important.

Reply:
I really like this idea. Thank you.
Done!